shared-libs/view-map-utils/src/view-map-utils.js

```diff
@@ -0,0 +1,157 @@
+'use strict';
+
+var emit = function() {},
+    viewMapStrings = {},
+    viewMapFns = {},
+    DEFAULT_DDOC_ID = 'default';
+
+//ensure V8 optimization
+var argumentsToArray = function () {
+  var args = [];
+  for (var i = this && this.skip || 0; i < arguments.length; i++) {
+    args.push(arguments[i]);
+  }
+  return args;
+};
+
+module.exports = {
+  reset: function (ddoc) {
+    if (ddoc) {
+      viewMapStrings[ddoc] = {};
+      viewMapFns[ddoc] = {};
+      return;
+    }
+
+    viewMapStrings = {};
+    viewMapFns = {};
+  },
+
+  loadViewMaps: function (ddoc) {
+    var ddocId = ddoc._id && ddoc._id.replace('_design/', '') || DEFAULT_DDOC_ID;
+    module.exports.reset(ddocId);
+    var viewNames = argumentsToArray.apply({ skip: 1 }, arguments);
+    viewNames.forEach(function(view) {
+      viewMapStrings[ddocId][view] = ddoc.views && ddoc.views[view] && ddoc.views[view].map || false;
+      viewMapFns[ddocId][view] = {};
+    });
+  },
+
+  getViewMapFn: function (ddocId, viewName, full) {
+    var COMMENT_REGEX = /\/\/.*/g,
+        SIGNATURE_REGEX = /emit\(/g,
+        NEW_LINE_REGEX = /\\n/g;
+
+    full = !!full;
+    if (viewMapFns[ddocId] && viewMapFns[ddocId][viewName] && viewMapFns[ddocId][viewName][full]) {
+      return viewMapFns[ddocId][viewName][full];
+    }
+
+    var fnString = module.exports.getViewMapString(ddocId, viewName);
+    if (!fnString) {
+      throw new Error('Requested view '+ ddocId + '/' + viewName + ' was not found');
+    }
+
+    fnString = fnString
+      .replace(NEW_LINE_REGEX, '\n')
+      .replace(COMMENT_REGEX, '')
+      .replace(SIGNATURE_REGEX, 'this.emit(')
+      .trim();
+
+    var fn = new Function('return ' + fnString)(); // jshint ignore:line
+
+    //support multiple `emit`s
+    var viewMapFn = function() {
+      var emitted = [];
+      var emit = function() {
+        return emitted.push(argumentsToArray.apply(null, arguments));
+      };
+      fn.apply({ emit: emit }, arguments);
+      return (full ? emitted : emitted[0]);
+    };
+    viewMapFns[ddocId][viewName][full] = viewMapFn;
+    return viewMapFn;
+  },
+
+  getViewMapString: function (ddocId, viewName) {
+    ddocId = ddocId || DEFAULT_DDOC_ID;
+    return viewMapStrings[ddocId] && viewMapStrings[ddocId][viewName] || module.exports.defaultViews[viewName];
+  },
+
+  defaultViews: {
+    docs_by_replication_key: function (doc) {
```

Comment:
Isn't this duplicated code?

Reply:
I see you've added a test to ensure they're identical to the ones in the ddoc at compile time - surely if we have this we don't need to worry about loading the views from the ddoc? We know it has to be the same as what's in `defaultViews`?